pkg/types/relayer.go

```diff
@@ -78,33 +78,34 @@ type NodeStatus struct {
 	State   string
 }
 
-// ChainService is a sub-interface that encapsulates the explicit interactions with a chain, rather than through a provider.
-type ChainService interface {
-	Service
-
+type ChainReader interface {
+	// NewContractReader returns a new ContractReader.
+	// The format of contractReaderConfig depends on the implementation.
+	NewContractReader(ctx context.Context, contractReaderConfig []byte) (ContractReader, error)
 	// LatestHead returns the latest head for the underlying chain.
 	LatestHead(ctx context.Context) (Head, error)
 	// GetChainStatus returns the ChainStatus for this Relayer.
 	GetChainStatus(ctx context.Context) (ChainStatus, error)
 	// ListNodeStatuses returns the status of RPC nodes.
 	ListNodeStatuses(ctx context.Context, pageSize int32, pageToken string) (stats []NodeStatus, nextPageToken string, total int, err error)
+}
+
+// ChainService is a sub-interface that encapsulates the explicit interactions with a chain, rather than through a provider.
+type ChainService interface {
+	Service
 	// Transact submits a transaction to transfer tokens.
 	// If balanceCheck is true, the balance will be checked before submitting.
 	Transact(ctx context.Context, from, to string, amount *big.Int, balanceCheck bool) error
 }
 
-// Relayer extends ChainService with providers for each product.
+// Relayer extends ChainService and ChainReader with providers for each product.
 type Relayer interface {
 	ChainService
-
+	ChainReader
```

Comment:
Oh, maybe we want:
```go
type Relayer interface {
	ChainReader
	ChainWriter
	...
}

type ChainService interface {
	Service
	ChainReader
	ChainWriter
}
type ChainWriter interface {
	// Transact submits a transaction to transfer tokens.
	// If balanceCheck is true, the balance will be checked before submitting.
	Transact(ctx context.Context, from, to string, amount *big.Int, balanceCheck bool) error

	// NewChainWriter returns a new ChainWriter.
	// The format of config depends on the implementation.
	NewChainWriter(ctx context.Context, config []byte) (ChainWriter, error)
}
```

Reply:
Yeah, but the ChainWriter has to be renamed to ContractWriter first